client/beefy/src/import.rs

```diff
@@ -0,0 +1,195 @@
+	Client as BeefyClient,
+};
+
+/// Checks the given header for a consensus digest signalling a beefy authority set change
+/// and extracts it.
+pub fn find_beefy_authority_set_change<B: BlockT>(
+	header: &B::Header,
+) -> Option<ValidatorSet<AuthorityId>> {
+	let id = OpaqueDigestItemId::Consensus(&BEEFY_ENGINE_ID);
+
+	let filter_log = |log: ConsensusLog<AuthorityId>| match log {
+		ConsensusLog::AuthoritiesChange(change) => Some(change),
+		_ => None,
+	};
+
+	// find the first consensus digest with the right ID which converts to
+	// the right kind of consensus log.
+	header.digest().convert_first(|l| l.try_to(id).and_then(filter_log))
+}
+
+/// BeefyBlockImport
+/// Wraps a type `inner` that implements [`BlockImport`]
+pub struct BeefyBlockImport<Backend, Block: BlockT, Client, I> {
+	client: Arc<Client>,
+	inner: I,
+	justification_sender: BeefyJustificationSender<NumberFor<Block>, Signature>,
+	_phantom: PhantomData<Backend>,
+}
+
+impl<BE, Block: BlockT, Client, I: Clone> Clone for BeefyBlockImport<BE, Block, Client, I> {
+	fn clone(&self) -> Self {
+		BeefyBlockImport {
+			client: self.client.clone(),
+			inner: self.inner.clone(),
+			justification_sender: self.justification_sender.clone(),
+			_phantom: PhantomData,
+		}
+	}
+}
+
+#[async_trait::async_trait]
+impl<BE, Block: BlockT, Client, I> BlockImport<Block> for BeefyBlockImport<BE, Block, Client, I>
+where
+	BE: Backend<Block>,
+	I: BlockImport<
+			Block,
+			Error = ConsensusError,
+			Transaction = sp_api::TransactionFor<Client, Block>,
+		> + Send
+		+ Sync,
+	Client: BeefyClient<Block, BE>,
+	Client::Api: BeefyApi<Block>,
+	for<'a> &'a Client:
+		BlockImport<Block, Error = ConsensusError, Transaction = TransactionFor<Client, Block>>,
+	TransactionFor<Client, Block>: 'static,
+{
+	type Error = ConsensusError;
+	type Transaction = TransactionFor<Client, Block>;
+
+	async fn import_block(
+		&mut self,
+		block: BlockImportParams<Block, Self::Transaction>,
+		new_cache: HashMap<well_known_cache_keys::Id, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		let hash = block.post_hash();
+		let number = *block.header.number();
+		let justifications = block.justifications.clone();
+		// If this block contains a beefy authority set change then it must have a beefy
+		// justification
+		let change = find_beefy_authority_set_change::<Block>(&block.header);
+		// Run inner block import
+		let import_result = self.inner.import_block(block, new_cache).await?;
+		// Try importing beefy justification
+		let beefy_justification =
+			justifications.and_then(|just| just.into_justification(BEEFY_ENGINE_ID));
+
+		if change.is_some() && beefy_justification.is_none() {
+			return Err(ConsensusError::ClientImport(
+				"Missing beefy justification in authority set change block".to_string(),
+			))
+		}
+		if let Some(beefy_justification) = beefy_justification {
+			self.import_justification(hash, number, (BEEFY_ENGINE_ID, beefy_justification))?;
+		}
```

Comment:
should we skip this if for old blocks (number < best_beefy)?

Reply:
I'm not yet sure how it's going to share state with the beefy-worker yet

Reply:
yes, nevermind, I realised that justification validation should happen anyway and old data can just be dropped on the worker side when we plug that in as well